metaflow/plugins/argo/event_decorators.py

```diff
@@ -0,0 +1,242 @@
+        logger,
+        echo,
+        options,
+        **kwarg
+    ):
+        self.attributes["trigger_set"] = None
+        self.attributes["error"] = None
+        self._option_values = options
+        (flows, events) = self._validate_inputs()
+        project_name = None
+        branch_name = None
+        # @project has already been evaluated
+        if "project_name" in current:
+            project_name = current.get("project_name")
+            branch_name = current.get("branch_name")
+        else:
+            if flow._flow_decorators.get("project"):
+                raise MetaflowException(
+                    "Move @project below @{}. ".format(self.name)
+                    + "Project namespacing must be applied before triggers are built."
+                )
+        self.attributes["trigger_set"] = TriggerSet(
+            project_name, branch_name, self.attributes["options"]
+        )
+        mappings = self.attributes.get("mappings")
+        is_aggregate = (len(flows) > 1) or (
+            events is not None and len(flows) + len(events) > 1
+        )
+        validated = validate_mappings(mappings, is_aggregate)
+        for flow in flows:
+            info = TriggerInfo(TriggerInfo.LIFECYCLE_EVENT)
+            info.name = flow
+            info.status = "succeeded"
+            if not is_aggregate and validated is not None:
+                info.mappings = {flow: validated}
+            else:
+                info.mappings = mappings
+            self.attributes["trigger_set"].append(info)
+        if events is not None:
+            for event in events:
+                info = TriggerInfo(TriggerInfo.USER_EVENT)
+                info.name = event
+                if not is_aggregate and validated is not None:
+                    info.mappings = {event: validated}
+                else:
+                    info.mappings = validated
+                self.attributes["trigger_set"].append(info)
+
+    def _validate_inputs(self):
+        flows = self.attributes.get("flows")
+        events = self.attributes.get("events")
+        if flows is not None and type(flows) != list:
+            raise bad_inputs_error("flows", "flow", self.name)
+        if events is not None and type(events) != list:
+            raise bad_inputs_error("events", "event", self.name)
+        if self.attributes["flow"] is not None:
+            flows.append(self.attributes["flow"])
+        if self.attributes["event"] is not None:
+            events.append(self.attributes["event"])
+        if len(flows) == 0 and len(events) == 0:
+            self._report_empty_inputs()
+        return (flows, events)
+
+    def _report_empty_inputs(self):
+        raise MetaflowException(
+            msg=(MISSING_INPUTS_MSG.format(self.name, "one flow or event name"))
+        )
+
+
+class TriggerOnFinishDecorator(TriggerOnDecorator):
+
+    name = "trigger_on_finish"
+    defaults = {"flow": None, "flows": [], "event": None, "options": dict()}
```

Comment:
@savingoyal : why do we have an event here? It should be just flows (or steps as mentioned earlier).

Reply:
@savingoyal : we also need cross branch/project support. This is fairly key to be able to "branch" off of a production flow and is used for debugging and experimentation.